src/java/org/apache/cassandra/service/disk/usage/DiskUsageMonitor.java

```diff
@@ -0,0 +1,229 @@
+
+    @VisibleForTesting
+    public DiskUsageMonitor()
+    {
+        this.dataDirectoriesSupplier = DiskUsageMonitor::dataDirectoriesGroupedByFileStore;
+    }
+
+    @VisibleForTesting
+    public DiskUsageMonitor(Supplier<Multimap<FileStore, Directories.DataDirectory>> dataDirectoriesSupplier)
+    {
+        this.dataDirectoriesSupplier = dataDirectoriesSupplier;
+    }
+
+    /**
+     * Start monitoring local disk usage and call notifier when local disk usage state changed.
+     */
+    public void start(Consumer<DiskUsageState> notifier)
+    {
+        // start the scheduler regardless guardrail is enabled, so we can enable it later without a restart
+        ScheduledExecutors.scheduledTasks.scheduleAtFixedRate(() -> {
+
+            if (!Guardrails.localDataDiskUsage.enabled(null))
+                return;
+
+            updateLocalState(getDiskUsage(), notifier);
+        }, 0, CassandraRelevantProperties.DISK_USAGE_MONITOR_INTERVAL_MS.getLong(), TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    public void updateLocalState(double usageRatio, Consumer<DiskUsageState> notifier)
+    {
+        double percentage = usageRatio * 100;
+        long percentageCeiling = (long) Math.ceil(percentage);
+
+        DiskUsageState state = getState(percentageCeiling);
+
+        Guardrails.localDataDiskUsage.guard(percentageCeiling, state.toString(), false, null);
+
+        // if state remains unchanged, no need to notify peers
+        if (state == localState)
+            return;
+
+        localState = state;
+        notifier.accept(state);
+    }
+
+    /**
+     * @return local node disk usage state
+     */
+    @VisibleForTesting
+    public DiskUsageState state()
+    {
+        return localState;
+    }
+
+    /**
+     * @return disk usage (including all memtable sizes) ratio
+     */
+    @VisibleForTesting
+    public double getDiskUsage()
+    {
+        // using BigDecimal to handle large file system
+        BigDecimal used = BigDecimal.ZERO; // space used by data directories
+        BigDecimal usable = BigDecimal.ZERO; // free space on disks
+
+        for (Map.Entry<FileStore, Collection<Directories.DataDirectory>> e : dataDirectoriesSupplier.get().asMap().entrySet())
+        {
+            usable = usable.add(BigDecimal.valueOf(usableSpace(e.getKey())));
+
+            for (Directories.DataDirectory dir : e.getValue())
+                used = used.add(BigDecimal.valueOf(dir.getRawSize()));
+        }
+
+        // The total disk size for data directories is the space that is actually used by those directories plus the
+        // free space on disk that might be used for storing those directories in the future.
+        BigDecimal total = used.add(usable);
+
+        // That total space can be limited by the config property data_disk_usage_max_disk_size.
+        DataStorageSpec diskUsageMaxSize = guardrailsConfigSupplier.get().getDataDiskUsageMaxDiskSize();
```

Comment:
The property `data_disk_usage_max_disk_size` is used to allow to set the disk usage limits as an absolute amount of space instead of relative to the free space on disk. Combined with the the percentages it's an alternative to have a pair of `data_disk_usage_size_warn_threshold`/`data_disk_usage_size_fail_threshold` properties, so we don't have two different limits. The use case would be, for example, a box with 60 TB local SSD volume running multiple Cassandra instances, or reserving disk space for other processes. So I'd say that this property is as important as the percentages themselves, since we are providing two alternative ways of defining the disk usage.

Reply:
I have tried to extend the comments about the new properties on yaml. As for `NEWS.txt`, the plan was to add a single entry about guardrails in general once we were closer to the release date. Now it could be that moment, so I have added that entry.

Reply:
LGTM